app/models/schema_registry.rb

```diff
@@ -2,37 +2,67 @@ module SchemaRegistry
   InvalidAvroSchemaError = Class.new(StandardError)
   IncompatibleAvroSchemaError = Class.new(StandardError)
 
-  def self.compatible?(compatibility, old_json, new_json)
-    direction = compatibility || Compatibility.global
-    old_schema = Schemas::Parse.call(old_json)
-    new_schema = Schemas::Parse.call(new_json)
-    case direction
-    when Compatibility::Constants::NONE
-      true
-    when Compatibility::Constants::BACKWARD
-      check(old_schema, new_schema)
-    when Compatibility::Constants::FORWARD
-      check(new_schema, old_schema)
-    when Compatibility::Constants::BOTH
-      check(old_schema, new_schema) && check(new_schema, old_schema)
+  extend self
+
+  def compatible?(new_json, version:)
+    compatibility = version.subject.config.try(:compatibility) || Compatibility.global
+
+    if Compatibility::Constants::TRANSITIVE_VALUES.include?(compatibility)
+      check_all_versions(compatibility, new_json, version.subject)
+    else
+      check_single_version(compatibility, version.schema.json, new_json)
     end
   end
 
-  def self.compatible!(compatibility, old_json, new_json)
-    unless compatible?(compatibility, old_json, new_json)
+  def compatible!(new_json, version:)
+    unless compatible?(new_json, version: version)
       raise IncompatibleAvroSchemaError
```

Comment:
No, just a boolean: http://docs.confluent.io/3.1.0/schema-registry/docs/api.html#id1

We could always add to the response. For the transitive checks here, I think the most obvious thing to add would be the first schema that was incompatible.

To get the reason why the schemas aren't compatible we'd have to plumb more information through in the avro gem. That's doable, but in the changes that I submitted I did not add exceptions that would allow the reason to bubble up.